Add actor blocking functionality with list table interface

obenland opened the pull request:
Fixes #418.

## Proposed changes:
* Implements a new actor blocking system similar to the existing following/followers functionality
* Uses post meta storage with `_activitypub_blocked_by` key on actor posts for scalable blocking
* Provides list table interface for managing blocked actors with add/unblock functionality
* Includes conditional menu registration that only shows if user has blocked actors
* Adds site-wide blocking support in Settings > ActivityPub tab
* Uses unified blocking API through `add_user_block`/`remove_user_block` methods

### Other information:

- [ ] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Go to Users > Blocked Actors (only appears if you have blocked actors)
* Try blocking an actor by entering `@username@domain.com` or profile URL
* Verify the actor appears in the blocked actors list
* Test unblocking individual actors and bulk unblock functionality
* Go to Settings > ActivityPub > Blocked Actors tab to test site-wide blocking
* Verify blocked actors cannot interact with your content

### Changelog entry

- [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [ ] Patch
- [x] Minor
- [ ] Major

#### Type

- [x] Added - for new features
- [ ] Changed - for changes in existing functionality
- [ ] Deprecated - for soon-to-be removed features
- [ ] Removed - for now removed features
- [ ] Fixed - for any bug fixes
- [ ] Security - in case of vulnerabilities

#### Message

Add actor blocking functionality with list table interface for managing blocked users and site-wide blocks

</details>

obenland committed:
Add comprehensive ActivityPub moderation system

Implements hierarchical blocking system (site-wide → user-specific → WordPress disallowed list) with three block types:
- Actors: Block specific ActivityPub users by URL
- Domains: Block entire instances by domain
- Keywords: Block content containing specific keywords

Features:
- Site-wide moderation for admins via Settings → ActivityPub
- User-specific blocks via user profiles
- Modern JavaScript with WordPress AJAX utilities
- Responsive UI matching WordPress standards
- Proper security with nonces and permission checks
- Integration with existing inbox controllers

obenland committed:
Add comprehensive PHPUnit tests for Moderation API

Implements thorough test coverage for the ActivityPub moderation system including:

- User block management (add/remove/get) for all three types (actors, domains, keywords)
- Site-wide block management with admin controls
- Activity blocking logic with hierarchical priority testing
- Edge cases: malformed data, invalid inputs, special characters, Unicode
- Complex scenarios: duplicate blocks, array re-indexing, WordPress fallback
- Boundary testing: extremely long values, invalid user IDs, empty data
- Integration testing: WordPress comment disallowed list fallback

All 22 tests pass with 91 assertions covering core functionality,
error handling, and real-world usage scenarios.

obenland committed:
Apply PHPCBF code formatting fixes

Automated code style fixes applied by PHP_CodeSniffer including:
- Consistent spacing and indentation
- WordPress coding standards compliance
- Array formatting and alignment
- Comment formatting improvements

No functional changes, only code style improvements.

obenland committed:
Improve moderation JavaScript UX and fix terminology

- Remove page reloads from AJAX operations for better UX
- Replace location.reload() with dynamic UI updates
- Add helper functions to manage DOM updates properly
- Fix table removal when last item is deleted
- Rename functions from 'block' to 'blockedTerm' to avoid confusion with WordPress blocks
- Fix issue where adding terms created tables in all sections
- Use specific selectors to target correct containers for each type

obenland committed:
Add code coverage annotations and refactor blocking logic

- Add @covers annotations to all test methods for proper coverage tracking
- Add @coversDefaultClass to test and main Moderation classes
- Refactor activity_is_blocked_site_wide and activity_is_blocked_for_user to eliminate code duplication
- Extract common blocking logic into private check_activity_against_blocks method
- Maintain full test coverage with all 22 tests passing

Improves code maintainability and test coverage visibility.

obenland committed:
Add changelog entry for blocking and moderation system

Documents the new comprehensive moderation features including:
- User-specific blocking controls for actors, domains, and keywords
- Site-wide moderation interface for administrators
- Integration with existing WordPress disallowed list
- Modern responsive UI with WordPress utilities

obenland committed:
Fix AJAX nonce verification order for better security

Move nonce verification to the beginning of all AJAX handlers before accessing any  data, as recommended by Copilot security review. This ensures proper nonce validation occurs before any other processing.

- ajax_add_user_block: Verify nonce first
- ajax_remove_user_block: Verify nonce first
- ajax_add_site_block: Verify nonce first
- ajax_remove_site_block: Verify nonce first, remove duplicate check

All tests continue to pass (22 tests, 91 assertions).

obenland committed:
Consolidate moderation admin functionality and fix AJAX parameter conflict

- Dissolve separate wp-admin/class-moderation.php into main class-admin.php
- Rename AJAX action to 'activitypub_moderation_settings' for clarity
- Fix AJAX parameter conflict by changing 'action' to 'operation'
- Remove separate moderation class initialization from main plugin file
- Integrate moderation scripts and handlers into unified Admin class
- Update JavaScript to use single nonce and new parameter names

The 'action' parameter was conflicting with WordPress's internal AJAX routing
mechanism, causing "Invalid context or action" errors. Using 'operation'
resolves this conflict.

Benefits:
- Reduces codebase by 70 lines while maintaining full functionality
- More descriptive AJAX action name clearly indicates purpose
- Simplified admin architecture with fewer files to maintain
- All 22 moderation tests continue to pass with 91 assertions

obenland committed:
Add moderation meta and settings for ActivityPub

Introduces user meta and site-wide settings for blocking ActivityPub actors, domains, and keywords. This enables more granular moderation capabilities for both individual users and the entire site.

obenland committed:
Align variable assignments in ajax_moderation_settings

Adjusted whitespace for variable assignments in the ajax_moderation_settings method to improve code readability and maintain consistent formatting.

obenland committed:
Fix test_add_user_block_valid_types by mocking remote actor response

obenland committed:
Remove explicit cap check

obenland committed:
Backslashit

obenland committed:
Remove actor block functionality from moderation settings

This commit removes the ability to block specific ActivityPub actors (users) by their actor URL from both site-wide and user moderation settings. The related UI elements, JavaScript logic, and server-side validation have been updated to only support domain and keyword blocks.

obenland committed:
Merge branch 'trunk' into add/block-lists

obenland committed:
Moderation: Refine UI with singular class names and table-only handling

obenland committed:
Use namespaced functions and improve moderation labels

Replaced calls to array_map and do_action with their namespaced versions for consistency and performance. Updated moderation-related labels and descriptions to use esc_html__ for proper escaping. Renamed Moderation::activity_is_blocked to Moderation::activity_is_blocked_for_user for clarity.

obenland committed:
Refactor moderation to remove actor-level blocking

Actor-level block functionality has been removed from the Moderation class and related tests. Blocking is now only supported for domains and keywords, simplifying the blocklist logic and data structures. Tests and settings have been updated to reflect this change, and duplicate domain entries are now prevented via array_unique in sanitization.

obenland committed:
Add actor blocking functionality with list table interface

Implements a new actor blocking system similar to following/followers with:
- Post meta storage using _activitypub_blocked_by key on actor posts
- List table interface for managing blocked actors with add/unblock functionality
- Conditional menu registration (only shows if user has blocked actors)
- Site-wide blocking support in Settings > ActivityPub tab
- Unified blocking API through add_user_block/remove_user_block methods
- Full admin interface with templates, screen options, and action handlers

obenland committed:
Merge branch 'trunk' into feature/new-actor-blocks

obenland committed:
Fix merge stuff

obenland committed:
Backslashit

obenland committed:
Merge branch 'trunk' into feature/new-actor-blocks

obenland committed:
Add periods to admin method docblocks and fix translation call

Standardized docblock comments in class-admin.php by adding periods at the end of descriptions. Also updated the translation function call in class-settings.php to use the global namespace for __().

obenland committed:
Add hooks for user and site block/unblock actions

Introduced new do_action hooks: 'activitypub_add_user_block', 'activitypub_remove_user_block', 'activitypub_add_site_block', and 'activitypub_remove_site_block'. These hooks allow other plugins or custom code to respond when actors, domains, or keywords are blocked or unblocked by users or site-wide, improving extensibility.

obenland committed:
Remove blocked actors from following list on block

Adds a hook to remove an actor from a user's following list when the actor is blocked. Implements the remove_blocked_actors method in Following to handle this logic, ensuring users no longer follow actors they block.

matticbot committed:
Add changelog

obenland committed:
Refactor actor blocking and icon handling

Simplifies the block_actor method by removing an unnecessary variable and streamlining error handling. Updates icon assignment to use object_to_uri for consistency when displaying blocked actors.

obenland committed:
Update label for profile input for accessibility

Changed the label for the profile input field to use 'Profile link' and added the 'screen-reader-text' class to improve accessibility for screen readers.

obenland committed:
Clean post cache after block/unblock actions

Added calls to clean_post_cache after adding or deleting blocked actors to ensure cache consistency. Also updated get_user_blocks and get_site_blocks to return only the 'guid' of blocked actors using wp_list_pluck.

obenland committed:
Refactor actor blocking logic in Blocked_Actors table

Inlines and updates the logic for blocking actors by handling email and URL profiles directly in the request handler. Removes the private block_actor() method and uses Moderation::add_user_block() directly, improving clarity and maintainability.

obenland committed:
Refactor unfollow to use actor ID instead of post

Updated the unfollow logic in Following class to use actor ID retrieved by Actors::get_id_by_various instead of the actor post object. This streamlines the process and ensures consistency in how actors are referenced.

obenland committed:
Merge branch 'trunk' into feature/new-actor-blocks

obenland commented on includes/wp-admin/table/class-blocked-actors.php:
```diff
@@ -0,0 +1,406 @@
+						\esc_html( $actor->post_title )
+					)
+				);
+			}
```
Not suer what it saw here, it does return a URL.

obenland commented on includes/wp-admin/table/class-blocked-actors.php:
```diff
@@ -0,0 +1,406 @@
+					\sprintf(
+						'<a href="%s">%s</a>',
+						\esc_url( \add_query_arg( 'resource', $search ) ),
+						\esc_html( $actor->post_title )
```
AI misread the context. It's a WP_Post

pfefferle committed:
Merge branch 'trunk' into feature/new-actor-blocks

pfefferle commented on templates/blocked-actors-list.php:
```diff
@@ -0,0 +1,82 @@
+							<input type="hidden" name="action" value="block" />
+							<input name="activitypub-profile" id="activitypub-profile" type="text" value="<?php echo esc_attr( $_resource ); ?>" size="40" aria-required="true" class="<?php echo $_resource ? 'highlight' : ''; ?>" />
+						</div>
+						<?php submit_button( esc_attr__( 'Block Actor', 'activitypub' ) ); ?>
```
I think `Block` is fine and it better aligns with the `Follow` button on the Following site.

```suggestion
						<?php submit_button( esc_attr__( 'Block', 'activitypub' ) ); ?>
```

pfefferle commented on templates/blocked-actors-list.php:
```diff
@@ -0,0 +1,82 @@
+						<div class="form-field form-required">
+							<label for="activitypub-profile" class="screen-reader-text"><?php echo esc_html__( 'Profile link', 'activitypub' ); ?></label>
+							<input type="hidden" name="action" value="block" />
+							<input name="activitypub-profile" id="activitypub-profile" type="text" value="<?php echo esc_attr( $_resource ); ?>" size="40" aria-required="true" class="<?php echo $_resource ? 'highlight' : ''; ?>" />
```
The Following form has a label for the profile URI. I know you are not a fan of it, but we should be consistent, so either we should remove the label on the following form or add it to the block form.

obenland replied:
Fixed in [48d3211](https://github.com/Automattic/wordpress-activitypub/pull/2027/commits/48d321141ed6e97d5649f5aa21c1f0cbfa33975b).

pfefferle commented:
If I block someone I follow or who follows me, unblocking them doesn’t restore that previous connection—it only removes them from my block list. If I had been following them, I can choose to follow again, but if they were following me, that connection is lost. It could be helpful to have an option to restore a follow in case a block was made by mistake.

pfefferle commented:
I see why you’re using the `Moderation` class to manage the Blocked Actors, but it’s not consistent with the rest of the codebase, where we typically use collection classes to query lists. While this isn’t necessarily an issue right now (or for us directly), it could make the code harder to read and understand in the long run, especially for contributors.

This isn’t a blocker—I’m not sure what the best approach here would be—but I thought it was worth pointing out.

pfefferle commented:
FYI: I may consolidate all feedback into this PR, since #1935 is based on it. That makes it tricky to tell which UI changes come from which PR :)

pfefferle commented:
There’s an edge case where an incoming Activity might use the WebFinger handle as its Actor-ID. Should we address that in this PR, treat it as an improvement for later, or just wait until the case actually comes up?

pfefferle commented on includes/class-moderation.php:
```diff
@@ -95,11 +101,45 @@ public static function activity_is_blocked_for_user( $activity, $user_id ) {
 	 */
 	public static function add_user_block( $user_id, $type, $value ) {
 		switch ( $type ) {
+			case 'actor':
```
What is the reason to have one function for all blocks instead of a dedicated function for blocking Actors? This function can become quite big and confusing.

Aside from that, I think we should use constants for `$type`.

pfefferle replied:
It also might be an argument for having a dedicated `Actor_Blocks_Collection`!

obenland replied:
> What is the reason to have one function for all blocks

To keep the API consistent and easy. Happy to have a blocked actors collection handle the management under the hood, but `Moderation` should be the top level API for it

obenland replied:
Constants fixed in d2736afdedfc4d5c6d0e321a190b458ddcbb384a.

pfefferle replied:
the question is still, why a global function with a type, instead of dedicated functions? there is nearly no shared code between the actor blocks and the keyword blocks.

obenland replied:
oh, because it started out using the same saving mechanism in options and user meta, so they all shared boilerplate code for adding/removing/checking.

In a subsequent step the actor stuff was removed, once it became clear it should get a user list treatment and a different way of storing that information

obenland replied:
See https://github.com/Automattic/wordpress-activitypub/pull/2020/commits/15679ca5ce1e0474307509a62dca2f41403add53

pfefferle commented on includes/class-moderation.php:
```diff
@@ -119,14 +159,50 @@ public static function add_user_block( $user_id, $type, $value ) {
 	 */
 	public static function remove_user_block( $user_id, $type, $value ) {
 		switch ( $type ) {
+			case 'actor':
```
same as above!

obenland replied:
Fixed in d2736afdedfc4d5c6d0e321a190b458ddcbb384a.

pfefferle commented on includes/class-moderation.php:
```diff
@@ -256,4 +356,57 @@ private static function check_activity_against_blocks( $activity, $blocked_actor
+	 *      @type int        $total         Total number of blocked actors.
+	 *  }
+	 */
+	public static function get_blocked_actors_with_count( $user_id, $number = -1, $page = null, $args = array() ) {
```
Shouldn't these live in the Actor class instead?

It also might be an argument for having a dedicated `Actor_Blocks_Collection`!

obenland replied:
Fixed in [625e23f](https://github.com/Automattic/wordpress-activitypub/pull/2027/commits/625e23f820dbd0eea1b9c0ad8ae4e2aa423685e8).

obenland committed:
Update templates/blocked-actors-list.php

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

obenland committed:
Align Following input label with Block input

See https://github.com/Automattic/wordpress-activitypub/pull/2027#discussion_r2300180934

obenland committed:
Resolve non-URL actors via Webfinger before blocking

Updated Moderation to resolve actor IDs that are not URLs using Webfinger before checking against blocked actors. Added unit tests to verify that actors in webfinger format are correctly resolved and blocked if their resolved URL matches a blocked domain.

obenland commented:
> There’s an edge case where an incoming Activity might use the WebFinger handle as its Actor-ID. Should we address that in this PR, treat it as an improvement for later, or just wait until the case actually comes up?

Fixed in [3c43505](https://github.com/Automattic/wordpress-activitypub/pull/2027/commits/3c43505462e60d9699192609c5f2d2d7b93acb46).

obenland committed:
Refactor blocked actors into dedicated collection class

- Create new Blocked_Actors collection class for better code organization
- Move get_blocked_actors_with_count() and get_blocked_actors() methods from Moderation class
- Update all references to use the new collection methods
- Maintain consistent naming with other collection classes (Followers, Following)
- Improve separation of concerns by grouping blocked actor functionality

obenland committed:
Move actor blocking management to Blocked_Actors collection

- Move add_block() and remove_block() methods from Moderation to Blocked_Actors collection
- Update Moderation class to delegate actor operations to the collection
- Keep Moderation as the top-level API while improving code organization
- Maintain all existing functionality including action hooks and cache clearing
- Improve separation of concerns with actor-specific logic in dedicated collection

obenland committed:
Fix PHP coding standards violations

- Remove extra blank line after class closing brace in Moderation class
- Resolve class name conflict by aliasing Blocked_Actors collection import
- Use Blocked_Actors_Collection alias in admin table to avoid namespace conflicts
- All composer lint checks now pass without errors

obenland committed:
Replace magic strings with TYPE constants in moderation system

- Add TYPE_ACTOR, TYPE_DOMAIN, TYPE_KEYWORD constants
- Update all switch statements to use constants instead of strings
- Replace array keys in USER_META_KEYS and OPTION_KEYS with constants
- Update blocked actors table to use Moderation::TYPE_ACTOR constant
- Maintain string literals for plural array keys where constants don't fit

obenland committed:
Merge branch 'trunk' into feature/new-actor-blocks

obenland commented:
@pfefferle This should be ready for another look.

obenland commented on tests/includes/class-test-moderation.php:
```diff
@@ -31,7 +31,7 @@ public function set_up(): void {
 
 		// Create a test user.
-		$this->test_user_id = $this->factory->user->create(
+		$this->test_user_id = self::factory()->user->create(
```
`$this->factory` is deprecated.

obenland commented on includes/wp-admin/table/class-blocked-actors.php:
```diff
@@ -0,0 +1,407 @@
+	 * Constructor.
+	 */
+	public function __construct() {
+		if ( \get_current_screen()->id === 'settings_page_activitypub' ) {
```
It only gets loaded within an admin context.

pfefferle commented:
@obenland what about

> If I block someone I follow or who follows me, unblocking them doesn’t restore that previous connection—it only removes them from my block list. If I had been following them, I can choose to follow again, but if they were following me, that connection is lost. It could be helpful to have an option to restore a follow in case a block was made by mistake.

???

obenland commented:
Ah, sorry, I missed replying to that. 

Can that be a follow-up? I'm not even sure that's something that's desired or that we'd need to worry too much about.

obenland committed:
Followers: First pass at blocking accounts (#1935)

pfefferle commented:
We should be at least sure about it, before we release it ;)

obenland commented:
I think for a v1 it's good enough to not support automatic re-following.

A lot of assumptions here, just as a heads up:
Blocking an actor is a pretty serious action that is probably not done lightly and relatively rare. Unblocking a blocked actor is probably even rarer, since there'll have been strong reasons to block in the first place and they would not have to be in place anymore. And then I'm not sure I'd consider unblocking the same as "everything is fine and normal again" and go back to the way it was.

Blocking is a pretty disruptive and destructive action, I think it's fine for it to be reflected in that way for the time being.

pfefferle commented on includes/wp-admin/table/class-following.php:
```diff
@@ -123,11 +124,34 @@ public function process_action() {
 
 				$profile = \sanitize_text_field( \wp_unslash( $_REQUEST['activitypub-profile'] ) );
-				if ( ! \is_email( \ltrim( $profile, '@' ) ) && empty( \wp_parse_url( $profile, PHP_URL_SCHEME ) ) ) {
+				$profile = \trim( $profile, '@' );
+
+				if ( \filter_var( $profile, FILTER_VALIDATE_EMAIL ) ) {
+					$remote_actor = Webfinger::resolve( $profile );
+					if ( ! \is_wp_error( $remote_actor ) ) {
+						$profile = object_to_uri( $remote_actor );
+					}
+				} elseif ( empty( \wp_parse_url( $profile, PHP_URL_SCHEME ) ) ) {
 					// Add scheme if missing.
 					$profile = \esc_url_raw( 'https://' . \ltrim( $profile, '/' ) );
 				}
 
+				// Check if user has blocked the account.
+				if ( \in_array( $profile, Moderation::get_user_blocks( $this->user_id )['actors'], true ) ) {
+					/* translators: %s: Account profile that could not be followed */
+					\add_settings_error( 'activitypub', 'followed', \sprintf( \__( 'Unable to follow account &#8220;%s&#8221;. The account is blocked.', 'activitypub' ), \esc_html( $profile ) ) );
+					$redirect_to = \add_query_arg( 'resource', $profile, $redirect_to );
+					break;
+				}
```
can you file an issue for this part to be refactored in a `Moderation::is_actor_blocked( $webfinger_or_uri_or_object_or_array )`?

pfefferle replied:
The function should also have a possibility to check either user, site or both.

obenland replied:
Done! #2095.

pfefferle commented on includes/wp-admin/table/class-following.php:
```diff
@@ -123,11 +124,34 @@ public function process_action() {
+				if ( \filter_var( $profile, FILTER_VALIDATE_EMAIL ) ) {
+					$remote_actor = Webfinger::resolve( $profile );
+					if ( ! \is_wp_error( $remote_actor ) ) {
+						$profile = object_to_uri( $remote_actor );
```
With this change, the form input will always be replaced with the ID, even if you have used the URL or WebFinger to follow. Is that fine?

obenland replied:
Oh man, I forgot about that. Yes, I noticed that, too. It feels a bit jarring but I couldn't think of a much better way. Maybe keep the original input separate?

obenland replied:
Fixed in [6d0db54](https://github.com/Automattic/wordpress-activitypub/pull/2027/commits/6d0db5454184bf76304400246203f581989b88f0).

obenland committed:
Preserve original profile in follow error redirects

Store the original profile value before resolving email-based profiles to URIs. Use the original value in error redirects to ensure the user sees the input they provided, improving clarity in error messages.
